clouddrift/wavelet.py

```diff
@@ -0,0 +1,184 @@
+
+    if norm == "energy":
+        psizero0 = np.exp(-(om**ga))
+        psizero1 = np.exp(be * np.log(om) - om**ga)
+        psizero = np.where(be == 0, psizero0, psizero1)
+    elif norm == "bandpass":
+        psizero0 = 2 * np.exp(-(om**ga))
+        psizero1 = 2 * np.exp(-be * np.log(fo) + fo**ga + be * np.log(om) - om**ga)
+        psizero = np.where(be == 0, psizero0, psizero1)
+    else:
+        raise ValueError(
+            "Normalization option (norm) must be one of 'energy' or 'bandpass'."
+        )
+
+    # to do: psizero[0] = 0.5*psizero[0] but am not sure how
+    # to do: replace NaN with zeros in psizero
+    psi = psizero
+
+    # to be continued
+    return psi, psif
+
+
+def morsefreq(
+    ga: Union[np.ndarray, float],
+    be: Union[np.ndarray, float],
+) -> Tuple[float, float, float]:
+    """
+    Frequency measures for generalized Morse wavelets. This functions calculates
+    three different measures of the frequency of the lowest-order generalized Morse
+    wavelet specified by parameters gamma (ga) and beta (beta).
+
+    Note that all frequency quantities here are *radian* as in cos(omega t)and not
+    cyclic as in np.cos(2 np.pi f t).
+
+    For be=0, the "wavelet" becomes an analytic lowpass filter, and fm
+    is not defined in the usual way.  Instead, fm is defined as the point
+    at which the filter has decayed to one-half of its peak power.
+
+    For details see Lilly and Olhede (2009).  Higher-order properties of analytic
+    wavelets.  IEEE Trans. Sig. Proc., 57 (1), 146--160.
+
+    Parameters
+    ----------
+    ga: np.ndarray or float
+       Gamma parameter of the wavelet.
+    be: np.ndarray or float
+       Beta parameter of the wavelet.
+
+    Returns
+    -------
+    fm: np.ndarray
+        The modal or peak frequency.
+    fe: np.ndarray
+        The "energy" frequency.
+    fi: np.ndarray
+        The instantaneous frequency at the wavelet center.
+    """
+    # add test for type and shape in case of ndarray
+    fm = np.where(
+        be == 0, np.log(2) ** (1 / ga), np.exp((1 / ga) * (np.log(be) - np.log(ga)))
+    )
+
+    fe = 1 / (2 ** (1 / ga)) * _gamma((2 * be + 2) / ga) / _gamma((2 * be + 1) / ga)
+
+    fi = _gamma((be + 2) / ga) / _gamma((be + 1) / ga)
+
+    return fm, fe, fi
+
+
+def morseafun(
+    ga: Union[np.ndarray, float],
+    be: Union[np.ndarray, float],
+    k: Optional[np.int64] = 1,
+    norm: Optional[str] = "bandpass",
+) -> float:
+    # add test for type and shape in case of ndarray
+    if norm == "energy":
+        r = (2 * be + 1) / ga
+        a = (2 * np.pi * ga * (2**r) * np.exp(lgamma(k) - lgamma(k + r + 1))) ** 0.5
+    elif norm == "bandpass":
+        om, _, _ = morsefreq(ga, be)
+        a = np.where(be == 0, 2, 2 / (np.exp(be * np.log(om) - om**ga)))
+    else:
+        raise ValueError(
+            "Normalization option (norm) must be one of 'energy' or 'bandpass'."
+        )
+
+    return a
+
+
+def _gamma(
```

Comment:
Checkout SciPy gamma and its log counterpart that accepts array-likes: https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.gamma.html#scipy-special-gamma

Reply:
I saw that but i questioned if it was worth it to have the whole scipy depency just for this?

Reply:
100%. I can explain in more detail on our call, but there are several benefits that outweigh the cost of increased package size.